File: Programmers/pro_60058_괄호변환.py
def isBalanced(p):
    cnt=0
    for i in p:
        if(i=='('):
            cnt+=1
        elif(i==')'):
        	cnt-=1
        if(cnt<0):
            return False
    return True
def stepTwo(p):
    cnt=0
    for i in range(len(p)):
        if(p[i]=='('):
            cnt+=1
        elif(p[i]==')'):
        	cnt-=1
        if (cnt==0):
            u=p[:i+1]
            v=p[i+1:]
            break
    return u,v
#3.
def solution(p):
	answer=''
    #1.
	if(p==''):
		return p
    #2.
	u,v=stepTwo(p)
	if(isBalanced(u)):
		return u+solution(v)
    #4.
	else:
		temp = '('
		temp +=solution(v)
		temp +=')'
		for i in range(1,len(u)-1):
			if(u[i]=='('):
				temp += ')'
			else:
				temp +='('        
		return temp
	return answer

File: Programmers/pro_60058_괄호변환.js
function solution(p) {
    if (p == '')
        return ''

    const isBalanced = (w) => {
        let [i, check] = [0, 0]
        for (; i < w.length; i++) {
            w[i] == '(' ? check += 1 : check -= 1
            if (check === 0)
                break
        }
        return [w.slice(0, i + 1), w.slice(i + 1, w.length)]
    }

    const isRight = (w) => {
        let check = 0
        for (let i = 0; i < w.length; i++) {
            w[i] == '(' ? check += 1 : check -= 1
            if (check < 0)
                return false
        }
        return check.length ? false : true
    }

    let answer = '';

    const getRight = (w) => {
        if (w == "")
            return ""
        let [u, v] = isBalanced(w)
        if (isRight(u))
            return u + getRight(v)
        let temp = "("
        temp += getRight(v) + ")"
        temp += [...u.slice(1, -1)].map((v) => v == "(" ? ")" : "(").join("")
        return temp
    }
    return getRight(p);
}